target/linux/ath79/dts/qca9531_compex_wpj531.dts

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+/dts-v1/;
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+
+#include "qca953x.dtsi"
+
+/ {
+	compatible = "compex,wpj531", "qca,qca9531";
+	model = "Compex WPJ531 (16MB flash)";
+
+	chosen {
+		bootargs = "console=ttyS0,115200n8";
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		sig1 {
+			label = "wpj531:red:sig1";
+			// TODO ar71xx says 13, hardware guide says 12
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		};
+
+		sig2 {
+			label = "wpj531:yellow:sig2";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+		};
+
+		sig3 {
+			label = "wpj531:green:sig3";
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
+
+		sig4 {
+			label = "wpj531:green:sig4";
+			gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
+		};
+
+		// TODO buzzer gpio 4 active high
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "Reset button";
+			linux,code = <KEY_RESTART>;
+			gpios = <&gpio 17 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&uart {
+	status = "okay";
+};
+
+&spi {
+	status = "okay";
+
+	flash@0 {
+		compatible = "jedec,spi-nor";
+		reg = <0>;
+		spi-max-frequency = <10000000>;
+
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			uboot:	partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x030000>;
+				read-only;
+			};
+
+			partition@30000 {
+				label = "firmware";
+				reg = <0x030000 0xfc0000>;
+				compatible = "denx,uimage";
+			};
+
+			art: partition@ff0000 {
+				label = "art";
+				reg = <0xff0000 0x010000>;
+				read-only;
+			};
+		};
+	};
+};
+
+&eth1 {
+	// TODO ar71xx loads MAC address from KSEG1ADDR(0x1f02e000) + offset
```

Comment:
i need help figuring out how to get these MAC addresses into the kernel

Reply:
That's what `mtd-mac-address` is for. Just have a look at the other devices' DTSes.
Instead of giving the full address, you specify a partition there (by it's node name) and then only provide the position based on the beginning of the partition.
For numeric increment of the address, there is the `mtd-mac-address-increment`, which should also be found easily in other DTSes.

Reply:
Looks like this might be
`mtd-mac-address = <&uboot 0x2e000>;`
(In case the uboot partition is really that big and not hiding a config partition or similar)

Reply:
oh, the offset:
eth0:
mtd-mac-address = <&uboot 0x2e010>;

eth1:
mtd-mac-address = <&uboot 0x2e018>;

Still, the addresses look a little bit too much "in the middle" of the partition.

Reply:
thank you very much :+1: 
i grep'd through the mtds - starting at 1 and not a 0 - and assumed it was calculated by uboot.

the stock u-boot kernel params include this line: `mtdparts=ath-nor0:256k(u-boot),64k(u-boot-env),6336k(rootfs),1408k(uImage),8256k(mib0),64k(ART)`

but the section doesn't look like a u-boot-env to me (and the other sizes don't look right either)
```
*
0002e000  20 13 04 17 00 00 00 00  00 00 00 00 00 00 00 00  | ...............|
0002e010  04 f0 21 38 71 cb 8e fd  04 f0 21 38 71 cc 4a cf  |..!8q.....!8q.J.|
0002e020  04 f0 21 38 71 cd f3 b5  04 f0 21 38 71 ce 3f 4c  |..!8q.....!8q.?L|
0002e030  00 00 00 00 00 00 55 5f  00 00 00 00 00 00 90 84  |......U_........|
0002e040  00 00 00 00 00 00 5a c6  00 00 00 00 00 00 b1 d4  |......Z.........|
0002e050  11 f6 06 93 11 f6 06 93  00 00 00 03 00 00 00 00  |................|
0002e060  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0002e0f0  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
0002e800  0f f9 99 dc a8 ed 27 aa  4f 5d 08 ac d3 18 71 02  |......'.O]....q.|
0002e810  d4 1f 1d 0f 14 af d2 84  52 9d ac d9 02 2b 24 6f  |........R....+$o|
0002e820  94 90 d6 bb e7 a2 d1 c9  c4 19 6f 88 e6 fe aa b4  |..........o.....|
0002e830  a1 51 75 d5 5d 2a a2 f3  b7 55 49 b6 52 48 4b ec  |.Qu.]*...UI.RHK.|
0002e840  ff ff ff ff ff ff ff ff  ff ff ff ff ff ff ff ff  |................|
*
00030000
```

Reply:
Hm, looks like you have FOUR MAC addresses for this device:
<&uboot 0x2e010>
<&uboot 0x2e018>
<&uboot 0x2e020>
<&uboot 0x2e028>

What are the other two used for? (ending cd and ce?) One might be WAN, the other might be WiFi, maybe it's a duplicate? Does your Wifi get the correct address already from calibration data? Which is it?

Despite, stock u-boot tells you 256k u-boot, that would be 0x40000 size for u-boot (you have 0x30000 in DTS!)
uboot-env would then be starting at 0x40000 and have a size of 0x50000. If its specified in the original firmware, I would definitely stick to that. Firmware partition then narrow down to 16000k, 0xfa0000 in size. Altogether will be 16*1024*1024=0x1000000 then, which sounds reasonable.
However, don't know what the mib0 partition is for and whether can just overwrite it.

Reply:
i've installed a stock ROM and couldn't find the last two mac addresses anywhere.
i'll have to get an untouched board to see what was on there before.

Reply:
Checking on an untouched board would be interesting.
However, this wouldn't be the first time we find unused MAC addresses...